### ISSUE TITLE
Icon layer icons thrash around when zooming in fully

### ISSUE THREAD

**Issue by ajfickas** [uber/deck.gl#647]
To reproduce:
1. Visit https://uber.github.io/deck.gl/#/documentation/layer-catalog/icon-layer.
2. Zoom all the way in on an icon.
3. Observe the icon's location jumps around significantly starting around zoom level 15 or so, appearing to jump around more as the map is zoomed in further.

I have reproduced the issue in Chrome 58.0.3029.110 (64-bit) and Safari 10.1 (12603.1.30.0.34) on macOs Sierra (10.12.4).

Is this a known issue? Any idea on the root cause and a possible fix or workaround?

**Comment by ibgreen** [comment]
> Is this a known issue? Any idea on the root cause and a possible fix or workaround?

@ajfickas Short answer: Click the 'fp64' checkbox and try again

Long answer: Yes, it is related to limited precision in 32 bit floating point values used to encode lng/lat. It starts to become apparent around meter resolutions, when you zoom in on city blocks and beyond.

deck.gl offers two solutions. Layers have an `fp64` prop that enables 64 bit GPU calculations. This solves the problem at the cost of some performance. You probably won't notice the difference unless you have a lot of icons (not sure where the limit is for IconLayer, maybe >100K).

The other solution is to use METER_OFFSET projection mode. You'd convert your icon positions to meter offsets from a center lng/lat and add layer props as follows:
```
              projectionMode: COORDINATE_SYSTEM.METERS,
              positionOrigin: [center.longitude, center.latitude],
```
This can require more work but is ultimately more performant.

**Comment by ajfickas** [comment]
@ibgreen Thanks for the quick response.

I checked the `fp64` checkbox, and I see no clear difference; the pins still jump around. Is there something else required to get the icons to render smoothly at meter resolutions in the example?

I see no clear difference when I pass `fp64: true` to the `IconLayer` in my code as well.

I'm experimenting with how to render using the `METER_OFFSET` projection mode in the meantime; however, the `fp64` prop sounds like the better solution for our use case since we're rendering << 100K icons at a time.

**Comment by shaojingli** [comment]
@ajfickas Hi Andrew

This means fp64 is not working on your computer, which warrants an investigation. Could you kindly share some details of the computer you are using, the GPU model and the OS it's running on? If you are using a mac, you can get it from "About This Mac..." from the apple menu on the top left corner.

**Comment by ajfickas** [comment]
@shaojingli Sure, here's the computer I'm running it on:

![screen shot 2017-05-22 at 3 00 03 pm](https://cloud.githubusercontent.com/assets/1221515/26324091/616653a4-3eff-11e7-9498-5bbcfae67d8e.png)

Let me know if you need anything else.

**Comment by shaojingli** [comment]
Hi @ajfickas can you pull down the latest deck.gl and run `npm run test-fp64`? You should see a browser window full of test results. Can you share those results to me? It could help pinpoint the issue.

**Comment by ajfickas** [comment]
@shaojingli Here you go:

[deck-gl-test-fp64-results.txt](https://github.com/uber/deck.gl/files/1056246/deck-gl-test-fp64-results.txt)

Run off `master` at 7b2045adeefcda9b1976769a5f95ae315ccb30ba.

**Comment by shaojingli** [comment]
@ajfickas thanks. It seems that fp64 indeed failed on your machine... all fp64 calculations lose accuracy and are carried out as 32-bit.... I'll try to find a similar machine to debug and keep you updated.

**Comment by ajfickas** [comment]
Thanks @shaojingli, appreciate it!

**Comment by ibgreen** [comment]
@shaojingli I now have a new MacBook albeit with a discrete GPU. But I can try to disable it and run the tests on my Intel GPU. But the fp64 test script no longer works...

**Comment by 1chandu** [comment]
This still needs to be fixed for newer Intel GPUs.

**Comment by 1chandu** [comment]
@shaojingli 
Some improvement has been made by @ib 's PR, but still not completely fixed. Given this affects only set of GPUs, it is a non blocker, we can fix this in 5.0 patch, moving to 6.0 for now.

**Comment by ib** [comment]
Hi, here is __ib__. Could you please use @ibgreen instead of @ib to address __ibgreen__, so that the right person will be notified. Thank you very much.

**Comment by ibgreen** [comment]
This should finally be fixed in 5.0.2.

**Comment by jianhuang01** [comment]
open for now until 5.0.2 is published.
